torch/csrc/jit/codegen/cuda/lower2device.cpp

```diff
@@ -349,12 +351,62 @@ void GpuLower::lower() {
   const auto conditional_loops =
       generateConditionalFromPredicate(fusion_, indexed_loops);
 
+  glfdc::ExpressionBuilder builder;
+  kir::ExprBuilderVisitor builder_visitor(builder, *kernel_);
+
   // Insert fake zero updates to make sure nvrtc doesn't blow out register use
   // on index and predicate reuse
   const auto register_adjusted = insertMagicZero(conditional_loops);
 
   // We now have the lowered expressions, finalize the kernel IR
-  kernel_->finalize(register_adjusted);
+  kernel_->finalize(register_adjusted, builder_visitor);
+
+  memoizeParallelIterExtents(builder_visitor);
+
+  kernel_->summary().memoized_dag = builder_visitor.takeDAG();
+  kernel_->summary().expr_reuses = builder_visitor.reuses();
+}
+
+void GpuLower::memoizeParallelIterExtents(kir::ExprBuilderVisitor& visitor) {
+  // Lets collect all IterDomains that are bound to a thread binding
+  std::vector<TensorView*> used_tvs;
+
+  // XXX: COPIED for setUsedTVs;
+  auto used_vals = fusion_->usedMathVals();
+  for (auto val : used_vals) {
+    if (val->getValType().value() == ValType::TensorView) {
+      used_tvs.push_back(val->as<TensorView>());
+    }
+  }
+
+  std::unordered_map<
+      ParallelType,
+      std::vector<std::pair<const kir::Val*, glfdc::ExprEvaluator>>,
+      TypeHash>
+      parallel_iter_extents;
+  for (auto tv : used_tvs) {
+    for (auto id : tv->domain()->domain()) {
+      if (id->isThread() && !id->isBroadcast()) {
+        // TODO(kir): we should rewrite this logic based on the Kernel object
+        auto kir_extent = this->lowerValue(id->extent());
+        const auto it = parallel_iter_extents.find(id->getParallelType());
+        auto expr = visitor.build(kir_extent);
+        TORCH_INTERNAL_ASSERT(expr.has_value());
+        if (it != parallel_iter_extents.end()) {
+          it->second.emplace_back(kir_extent, expr.value());
+        } else {
+          std::vector<std::pair<const kir::Val*, glfdc::ExprEvaluator>> extents(
+              1,
+              std::pair<const kir::Val*, glfdc::ExprEvaluator>(
+                  kir_extent, glfdc::ExprEvaluator(expr.value())));
+          parallel_iter_extents[id->getParallelType()] = std::move(extents);
+        }
+      }
+    }
+  }
```

Comment:
This seems to make sense to me, but I wonder why it was previously done at the launch time. Even without the new expression evaluator, we could just cache `kir::Val*` for each parallel type so that at the launch time we wouldn't need to analyze the kernel every time. 

@jjsjann123, do you have any concern here? Specifically, this part of the code is now moved at the lowering time:

https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/executor.cpp#L339-L355

Reply:
Looks like the code is just moved to a different place but the logic follows the same.

IIRC, `computeLaunchParams` is used in kernel segmentation as well, but I don't think their parallel type binding would change so @naoyam's suggestion here for caching that sounds valid. cc'ing @shmsong

Reply:
Makes sense. Caching the `Val`'s here will save one usedTV traversal on the runtime. Segmenter doesn't look at the parallel types and launch parameters, or at least not yet. But I can add caching of the parallelized extents right after each segment is scheduled if that sounds like a good idea. 

Also another thing to note is that launch parameters are re-computed only if we have an input miss but we have a heuristic hit, which is the second level hit in our caching system. The fastest code path where we have exact input dimension match wouldn't need to compute launch params. 

On a separate future thread when we revisit run time. We probably need more profiling effort to see how important is the speed of the heuristic hit, and how important `computeLaunchParams` is in heuristic hit delay.

Reply:
Thanks @jjsjann123 and @shmsong. I feel more confident with this change.